File: src/rosbot_cpp_examples/src/simple_subscriber.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

using std::placeholders::_1;

// SimpleSubscriber listens for messages on the "chatter" topic.
class SimpleSubscriber : public rclcpp::Node
{
public:
    // Construct the node and set up the subscription.
    SimpleSubscriber(): Node("simple_subscriber")
    {
        // Subscribe to the "chatter" topic with queue depth 10.
        subscriber_ = create_subscription<std_msgs::msg::String>(
            "chatter", 10, std::bind(&SimpleSubscriber::topic_callback, this, _1));

        // Log that the node is ready.
        RCLCPP_INFO(get_logger(), "Simple Subscriber node has been started.");
    }

private:
    // ROS 2 subscription for std_msgs/String messages.
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscriber_;

    // Handle each incoming message from the subscribed topic.
    void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
    {
        RCLCPP_INFO_STREAM(get_logger(), "I heard: " << msg->data);
    }
};

int main(int argc, char * argv[])
{
    // Initialize ROS 2, run the node, then shut down cleanly.
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleSubscriber>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/rosbot_py_examples/rosbot_py_examples/simple_subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class SimpleSubscriber(Node):
    """ROS 2 node that subscribes to greeting messages on /chatter."""

    def __init__(self):
        # Initialize the node with the ROS graph name `simple_subscriber`.
        super().__init__('simple_subscriber')

        # Create a String subscription:
        # - message type: std_msgs/String
        # - topic name: chatter
        # - callback: listener_callback
        # - queue depth: 10
        self.subscriber_ = self.create_subscription(
            String,
            'chatter',
            self.listener_callback,
            10
        )
        # Keep a reference to avoid an unused variable warning.
        self.subscriber_  # prevent unused variable warning

    def listener_callback(self, msg):
        # Log each received message to confirm subscription activity.
        self.get_logger().info('I heard: "%s"' % msg.data)

def main():
    # Standard ROS 2 lifecycle:
    # 1) initialize rclpy
    # 2) create and spin node (process callbacks)
    # 3) clean up node and shutdown rclpy on exit
    rclpy.init()
    simple_subscriber = SimpleSubscriber()
    rclpy.spin(simple_subscriber)
    simple_subscriber.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
